### PR TITLE
feat(tt): tip progress in indexeddb (local only)

### PR DESCRIPTION
**Pull request by joelhooks**
This uses the ✅ all over and shouldn't be approved without that being changed, so it's draft.

Using IndexedDB we can have a database of events. Currently only looking at the video complete event for tips. This could later be sync'd to database and/or prefer the database as needed.

Using react-query means it will cache for general display.

screenshot

gif

### CODE REVIEW

File: apps/total-typescript/src/hooks/use-mux-player.tsx [hunk 1]
```diff
@@ -33,11 +33,19 @@ type VideoProviderProps = {
   lesson: VideoResource
   path?: string
   muxPlayerRef: any
+  onEnded?: () => Promise<any>
```
Comment (joelhooks) [inline apps/total-typescript/src/hooks/use-mux-player.tsx:36]: wanted to be able to track in the local context, specifically to write to database

File: apps/total-typescript/src/hooks/use-tip-complete.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+})
+
+export type TipEvent = z.infer<typeof TipEventSchema>
+
+export const useTipComplete = (tipSlug: string) => {
+  const {getManyByIndex} = useIndexedDBStore('progress')
+  const {data: completionEvents, status} = useQuery(
+    ['completionEvents', tipSlug],
+    async () => {
+      const tipEvents = await getManyByIndex('lesson', tipSlug)
```
Comment (joelhooks) [inline apps/total-typescript/src/hooks/use-tip-complete.ts:19]: access indexeddb for all objects with 'lesson' property that matches the tip

File: apps/total-typescript/src/hooks/use-tip-complete.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      return z
+        .array(TipEventSchema)
+        .parse(tipEvents)
+        .filter((tipEvent) => {
+          return (
+            tipEvent.eventName === 'completed video' &&
+            tipEvent.module === 'tips'
```
Comment (joelhooks) [inline apps/total-typescript/src/hooks/use-tip-complete.ts:26]: as the hook name implies, we only care about tip completions

File: apps/total-typescript/src/pages/tips/index.tsx [hunk 4]
```diff
@@ -129,7 +131,10 @@ export const TipTeaser: React.FC<{tip: Tip}> = ({tip}) => {
             },
           }}
         >
-          <a className="hover:underline">{title}</a>
+          <a className="hover:underline">
+            {tipCompleted ? '✅ ' : ''}
```
Comment (joelhooks) [inline apps/total-typescript/src/pages/tips/index.tsx:135]: can sprinkle this around however you like, not context dependent

File: apps/total-typescript/src/templates/tip-template.tsx [hunk 5]
```diff
@@ -44,8 +50,22 @@ const TipTemplate: React.FC<TipPageProps> = ({tip, tips}) => {
...
   })
 
+  const handleVideoEnded = async () => {
+    await add({
+      eventName: 'completed video',
+      module: 'tips',
+      lesson: tip.slug.current,
+      createdOn: new Date(),
+    }).then(console.debug)
+  }
```
Comment (joelhooks) [inline apps/total-typescript/src/templates/tip-template.tsx:60]: we are adding the dip here

File: apps/total-typescript/src/templates/tip-template.tsx [hunk 6]
```diff
@@ -212,63 +231,81 @@ const TipOverlay: React.FC<{tips: Tip[]}> = ({tips}) => {
...
+          {take(shuffle(tips), 9).map((tip) => (
+            <VideoOverlayTipCard suggestedTip={tip} />
+          ))}
         </div>
       </div>
     </div>
   )
 }
 
+const VideoOverlayTipCard: React.FC<{suggestedTip: Tip}> = ({suggestedTip}) => {
```
Comment (joelhooks) [inline apps/total-typescript/src/templates/tip-template.tsx:243]: extracted this out so it's a FC instead of a "callback" so it can use hooks directly

**Review by vojtaholik — APPROVED** [review]

**Pull request merged**